ManagedEntities - Allow targeted reconciliation. Add hook parameter. Simplify logic.

colemanw opened the pull request:
Overview
----------------------------------------
Makes the ManagedEntity reconcile process more efficient and the code easier to read & understand.
Makes saving or reverting an Afform faster by only reconciling managed entities from that one module.

Before
-------

* `CRM_Core_ManagedEntities->reconcile()` reconciles entities in _all_ modules.
* `hook_managed(&$entities)` enumerates entities in _all_ modules.

After
-------

* `CRM_Core_ManagedEntities->reconcile($modules = NULL)` can optionally be limited to specific modules.
* `hook_managed(&$entities, $modules = NULL)` can optionally limit lookups to specific modules. 

    *All core extensions now support `hook_managed` with the `$modules` filter. Existing implementations of `hook_managed` are not required to abide by `$modules`, but doing so will improve performance. Example:*

    ```php
    function example_civicrm_managed(&$entities, ?array $modules = NULL) {
      if ($modules !== NULL && !in_array(E::LONG_NAME, $modules, TRUE)) {
        return;
      }
      // Otherwise, continue and add $entities.
    }
    ```

Technical Details
----------------------------------------
The `CRM_Core_ManagedEntities` class has been around for a while and gathered a bit of cruft in that time.
I found the code quite difficult to read and understand, as it was fetching and re-fetching records, sorting them and dividing them by module but ultimately discarding those divisions. Some records were getting processed twice. Here's what I did:

- Instead of fetching Managed records 3 times (for enabled, disabled and missing modules), now they are fetched only once and sorted by action (create, update, disable, delete).
- Instead of looping through each module and then processing some of its records (leaving gaps that have to be filled in later), loop through all the managed entities once and process them all.
- The `reconcile()` function now accepts a `$moduleName` param that will reconcile a single module, which fixes a fixme in Afform.
- Removed a lot of extra sorting that didn't serve any purpose.

colemanw committed:
REF - Refactor ManagedEntities with the ability to reconcile a single module

Reconciling a single module (e.g. Afform) is more efficient.
Further refactoring of ManagedEntities for code simplification and efficiency.
Instead of fetching Managed records 3 times (for enabled, disabled and missing modules),
now they are fetched only once and sorted into the correct category.

colemanw commented on ext/afform/core/Civi/Api4/Utils/AfformSaveTrait.php:
```diff
@@ -70,8 +72,7 @@ protected function writeRecord($item) {
     ) {
-      // FIXME: more targeted reconciliation
-      \CRM_Core_ManagedEntities::singleton()->reconcile();
+      \CRM_Core_ManagedEntities::singleton()->reconcile(E::LONG_NAME);
```
@totten this FIXME has been staring me down for a long time. Finally fixed!

totten replied:
Woot! That should help performance when editing afforms!

eileenmcnaughton commented on CRM/Core/ManagedEntities.php:
```diff
@@ -115,19 +115,14 @@ public function get($moduleName, $name) {
    *
-   * @param bool $ignoreUpgradeMode
-   *   Unused.
+   * @param str $moduleName
```
this should be string not str

totten commented:
jenkins, test this please

totten commented:
So I did read on the code before-patch and after-patch, and I agree this looks better.

Just to verbalize... it appears to me (*and this is post-hoc rationalization*) that `ManagedEntities.php` has made a transition in style:

* _In the original style_, it had qualitatively distinct stages that targeted different situations (`reconcileEnabledModules()`, `reconcileDisabledModules()`, `reconcileUnknownModules()`). Each stage ran separately -- using its own techniques to find interesting records and act upon them.
* _In the final style_, it has a _planner_  which builds a _todo list_. (ie `loadManagedEntityActions()` builds `$managedActions`.) Then it walks through the plan and executes each todo (`reconcileEntities()`).
* _In the transition_, it nominally used both styles. It first ran a general planner, and then it ran each of the stages -- and those stages cherry-picked todos from the plan (whichever todos seemed interesting to them). But this split makes it harder to sense the overall dataflow.

IMHO, the planner/todo-list approach is better than the 3-stage approach. eg I like that the planner is more function-y -- you get more ways to test+inspect the todo-list without modifying the DB. It may also be more amenable to re-sorting based on a dependency-graph.

This PR definitely makes it easier to read.

On `r-code-style` front, I'd be tempted to go a bit further. Personally, the names `$managedActions` and `loadManagedEntityActions(): void` are confusing. (`$managedActions` sounds equivalent to `ScheduledJobs` to me, and `loadBlahBlahActions` sounds like it's going to read an action-list from somewhere else. But it's really synthesizing the data.) I'd be more partial to `$todos` or `$tasks` or `$plan` or `createPlan($declarations): array` or `buildTaskList($declarations): array`. But that's bike-sheddy, and distinct from the other stuff in this PR.

There is a fair amount of test-coverage in `CRM_Core_ManagedEntitiesTest` and `api\v4\Entity\ManagedEntityTest`.

On the general conceptual level, I'm inclined toward merging.

On a more concrete QA level, there are also several smaller places where it twiddles types+names (eg DAO=>array) which I haven't really looked at. (I'm optimistic because those are mostly internal and because there's test-coverage. But I didn't have the general picture when I first skimmed them, and I should probably give them another read with this better sense of context.)

totten committed:
hook_managed - Accept option module filter

colemanw commented:
@totten thanks for taking the time to get your head round this, I think you summed it up well, and I think you're right the *before* state of the class is transitional, and this PR basically completes the transition.
My reading of your conclusion is basically "this is good but could go further." If that's the case then I suggest we merge this & follow-up with another PR to finish renaming the variables, etc.

eileenmcnaughton commented:
I also agree with the analysis - I think the intent of the transitional refactor was a step towards the plan & then do approach

totten commented on CRM/Core/ManagedEntities.php:
```diff
@@ -655,23 +534,41 @@ private function isActivationSupported(string $entity_type): bool {
+    // Ideally, given a $moduleName like 'org.foo.demo' we'd just call the function `demo_civicrm_managed()`
+    // But alas, that only works with old-style hooks and not the event dispatcher.
+    // So here we go loading declarations from every module whether we need them or not.
+    CRM_Utils_Hook::managed($declarations);
```
Comment responding to a comment... `CRM_Utils_Hook` dispatches to several different systems (Drupal, Joomla, WordPress, Civi extensions, and Symfony listeners). The problem isn't with any one system - it's that multiple-dispatch requires a "lowest common denominator" approach, and several systems (*Symfony+WordPress+Joomla*) don't support the trick you describe. 

If one doesn't need portability or backward-compatibility, then we could define a different interface for registering mgds (and specifically support this).

However, even within the current hook, I think this could be largely improved - eg add a second parameter:

https://github.com/civicrm/civicrm-core/commit/e452a1485a0bf5860bd68370800e9ba6b4064228

This doesn't quite stop the dispatch - but it can prevent unnecessary file-scans (which is probably 90% of the cost). Note that this draft changes the `mixin` -- so the optimization would roll-out to all core extensions (which already use the mixin) and to [future/migrated `civix` extensions](https://github.com/totten/civix/pull/233).

colemanw replied:
@totten I like that commit. I notice that you make it into an array, whereas this PR only considers a single `$module`. We should probably be consistent there. We could cast it to an array as soon as it gets passed into `reconcile()` and then it would always be multivalued.

totten replied:
Good idea @colemanw.

I can send up that commit as a follow-up PR.

totten replied:
Oops, nevermind ^^^. I see you cherry-picked it. :)

totten commented on CRM/Core/ManagedEntities.php:
```diff
@@ -600,16 +479,17 @@ protected function isModuleDisabled(string $module): bool {
-    foreach ($declarations as $name => &$declare) {
-      if (!array_key_exists('name', $declare)) {
-        $declare['name'] = $name;
+  protected function setDeclarations(array $declarations, $moduleName = NULL) {
```
(*Non-blocker/General observation*) There's a tension around the lifecycle of `$declarations`, which has behavioral and stylistic dimensions:

* (Behaviorally) Are `$declarations` ephemeral things (_loaded+destroyed each time you use `ManagedEntities`_) or are they retained for a while (_re-used in multiple calls to `ManagedEntities`_)? How long should they live in memory?
* (Stylistically) Do we prefer to use more stateless design (eg `cleanupDeclarations(array):array`) or more stateful design (eg `setDeclarations(array,string): void`)?

A couple thoughtlets that stand out to me:

* There are two `public` operations relying on this data (`reconcile()` + `revert()`). Both of them rebuild the `$declarations` before doing their work. This behavior seems reasonable to me (*though maybe that belief merits inspection*). In any case, it appears consistent before+after patch.
* AFAICS, the `$moduleName` is never actually passed in to `setDeclarations()`. I was concerned it would mean some subtle corruption of `$this->declarations`. But it doesn't matter - b/c `reconcile()`+`revert()` will rebuild `$this->declarations` whenever they're called.
* It seems to me that (following the other changes) there's now less need to store data in `$this`. In the older design, it retained `$this->declarations` so that each `reconcileFooModules()` could do a pass through the list -- in the newer design, the planner does a single pass over `$declarations`.

Anyway, this comment has as many questions as answers, and these don't have to be resolved now. The relevant methods are `protected function`s in a _de-facto_ `final class`, so we can continue to revise based on other analysis or discussion.

colemanw commented:
@totten I've rebased this PR to include your suggestion about variable/function names, made `$modules` multivalued, tweaked and added your additional commit.

totten commented:
@colemanw OK, cool. I'm flagging it "merge ready". I should give it some `r-run` (due diligence) since it's changed -- and if I do that today, then I'll merge it. (But if not, then it's probably fine as long as tests pass - and you can merge tomorrow.)

totten commented:
Did some light `r-run` in a few areas. These all worked as expected:

* Running `tools/mixin/bin/test-all`
* (Afform) Toggling dashlet support, and then checking that the dashlets are added/removed on the dashboard
* (General) Go through lifecycle of installing/uninstalling various extensions that rely on managed data (esp CiviGrant and afform_admin)

Promoting to "merge on pass".

colemanw commented:
Thanks @totten. Here's a followup because I realized that `ManagedEntities->reconcile()` is getting called from outside of core... because there's no API.

colemanw commented on CRM/Core/ManagedEntities.php:
```diff
@@ -600,16 +479,17 @@ protected function isModuleDisabled(string $module): bool {
-      if (!array_key_exists('name', $declare)) {
-        $declare['name'] = $name;
+  protected function setDeclarations(array $declarations, $moduleName = NULL) {
+    $this->declarations = [];
```
@totten I have been pushing in the direction of ephemeral. A few versions ago I hit some bugs caused by them hanging around from one reconcile to the next and they were solved by the addition of this line.
I would be happy to push it further in that direction and make them local variables passed between functions instead of class variables.

colemanw replied:
https://github.com/civicrm/civicrm-core/pull/23251

totten committed:
ManagedEntitiesTest - Add coverage for filtering by module

totten commented:
> Thanks @totten. Here's a followup because I realized that ManagedEntities->reconcile() is getting called from outside of core... because there's no API.

I assume that's a reference to #23243... Commented over there.

Pushed up a little more test-coverage here.

Expanded description to highlight changes to two interfaces (`reconcile()` and `hook_managed`).

Still `merge on pass`.
